### PR TITLE
Emit more helpful content in producers section

### PR DESCRIPTION
**Pull request by jeffcharles**

### CODE REVIEW

File: crates/cli/tests/dynamic_linking_test.rs [hunk 1]
```diff
@@ -39,6 +39,22 @@ pub fn test_dynamic_linking() -> Result<()> {
...
+fn test_producers_section_present() -> Result<()> {
+    let js_wasm = create_dynamically_linked_wasm_module("console.log(42)")?;
+    let producers_section = wasmparser::Parser::new(0)
+        .parse_all(&js_wasm)
+        .find_map(|payload| match payload {
+            Ok(wasmparser::Payload::CustomSection(c)) if c.name() == "producers" => Some(c.data()),
+            _ => None,
+        })
+        .unwrap();
+    let producers_string = str::from_utf8(producers_section)?;
```
Comment (jeffcharles) [inline crates/cli/tests/dynamic_linking_test.rs:52]: The data section contents isn't a UTF-8 string but contains a series of UTF-8 strings. But the non-UTF8 parts are just unsigned LEB128 values that are `1` or `2` so this works as a very easy way to parse the byte array so we can do a quick inspection at test time.

**Review by saulecabrera — COMMENTED** [review]
Generally looks good; I have a couple of high level thoughts that I'd like to work through before landing this:

1. This is the second module level transformation that we have in Javy and as such, I think that we should be asserting that that the produced modules are valid. As far as I know we are not doing that anywhere; we should call `assert (wasmparser::validate(&wasm_bytes).is_ok());` after all the transformations are done.

2. The transformations pass needs a bit of consolidation IMO: right now we have two different ways, using two different dependencies of doing module level transformations: one with parity-wasm for the statically linked module and one with `wasmparser`/`wasm-encoder` for the dynamic modules. For the statically linked modules this poses a problem: with this change we are parsing and encoding the module twice, which is going to slow down the process of generating the final Wasm module. I think that we can re-structure the code in such a way that  parsing / encoding modules  happens only once, _after_ all the transformations are done. I'm thinking that we could have a `transformations` module that exposes the functions for the transformations that we want to do (e.g. `add_source_code_custom_section`, `add_producers_section`). The current approach feels a bit scattered, the way in which the transformation functions are placed. 

3. As noted in one of my comments, we should get rid of the `parity-wasm` dependency.

File: crates/cli/Cargo.toml [hunk 2]
```diff
@@ -27,6 +27,8 @@ wasmprinter = { version = "0.2.45", optional = true }
 wasmtime = { workspace = true }
 wasmtime-wasi = { workspace = true }
 wasi-common = { workspace = true }
+wasmparser = "0.101.0"
```
[the remark this answers is not in the document]
Comment (jeffcharles) [reply crates/cli/Cargo.toml:30]: I can do a separate refactoring PR to swap `parity-wasm` out with `wasm-encoder` and then rebase this change on top of that.

`wasmparser` doesn't allow random access directly but you can use it with `wasm-encoder` to read and encode all the sections into the new module.

Also potentially a separate refactoring PR to make the dynamic linking code use the same code as we would use for static linking to actually add the sections (that is also have the dynamic mode code, dump the module to disk without the custom sections, then re-read bytes from disk, parse the module, and add the sections).

**Comment by jeffcharles** [comment]
Opened https://github.com/Shopify/javy/pull/245 to do some preliminary refactoring so we can add this in a more clean way.

**Review by saulecabrera — APPROVED** [review]

**Pull request merged**